src/main/java/ch/njol/skript/patterns/TypePatternElement.java

```diff
@@ -0,0 +1,200 @@
+/**
+ *   This file is part of Skript.
+ *
+ *  Skript is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  Skript is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with Skript.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Copyright Peter Güttinger, SkriptLang team and contributors
+ */
+package ch.njol.skript.patterns;
+
+import ch.njol.skript.Skript;
+import ch.njol.skript.classes.ClassInfo;
+import ch.njol.skript.lang.Expression;
+import ch.njol.skript.lang.Literal;
+import ch.njol.skript.lang.SkriptParser;
+import ch.njol.skript.lang.SkriptParser.ExprInfo;
+import ch.njol.skript.lang.parser.ParserInstance;
+import ch.njol.skript.log.ErrorQuality;
+import ch.njol.skript.log.ParseLogHandler;
+import ch.njol.skript.log.SkriptLogger;
+import ch.njol.skript.registrations.Classes;
+import ch.njol.skript.util.Utils;
+import ch.njol.util.Kleenean;
+import ch.njol.util.NonNullPair;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * A {@link PatternElement} that contains a type to be matched with an expressions, for example {@code %number%}.
+ */
+public class TypePatternElement extends PatternElement {
+
+	private final ClassInfo<?>[] classes;
+	private final boolean[] isPlural;
+	private final boolean isNullable;
+	private final int flagMask;
+	private final int time;
+
+	private final int expressionIndex;
+
+	public TypePatternElement(ClassInfo<?>[] classes, boolean[] isPlural, boolean isNullable, int flagMask, int time, int expressionIndex) {
+		this.classes = classes;
+		this.isPlural = isPlural;
+		this.isNullable = isNullable;
+		this.flagMask = flagMask;
+		this.time = time;
+		this.expressionIndex = expressionIndex;
+	}
+
+	public static TypePatternElement fromString(String s, int expressionIndex) {
+		boolean isNullable = s.startsWith("-");
+		if (isNullable)
+			s = s.substring(1);
+
+		int flagMask = ~0;
+		if (s.startsWith("*")) {
+			s = s.substring(1);
+			flagMask &= ~SkriptParser.PARSE_EXPRESSIONS;
+		} else if (s.startsWith("~")) {
+			s = s.substring(1);
+			flagMask &= ~SkriptParser.PARSE_LITERALS;
+		}
+
+		if (!isNullable) {
+			isNullable = s.startsWith("-");
+			if (isNullable)
+				s = s.substring(1);
+		}
+
+		int time = 0;
+		int timeStart = s.indexOf("@");
+		if (timeStart != -1) {
+			time = Integer.parseInt(s.substring(timeStart + 1));
+			s = "" + s.substring(0, timeStart);
```

Comment:
Is the `"" + ` necessary?

Reply:
Nope, I probably copied this part from the old parser